Do not hit main thread when using CacheStorageConnection from a worker

youennf opened the pull request:
#### 413b6df42e717eff645d3c4ca47a332ddb64edfe
<pre>
Do not hit main thread when using CacheStorageConnection from a worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=274967">https://bugs.webkit.org/show_bug.cgi?id=274967</a>
<a href="https://rdar.apple.com/129062867">rdar://129062867</a>

Reviewed by Sihui Liu.

We make WebCacheStorageConnection thread safe to be called.
To do so, we add locks for getting the connection and for the hash maps.
To keep consistency between connection and hash maps, as was ensured before this patch, we lock both map lock and connection lock when updating the map.

We block the worker thread to get the connection and then we can use it directly from the worker.
In case of network process crash, we might have to block again the worker on main thread.

Covered by existing tests.

* Source/WebCore/Modules/cache/WindowOrWorkerGlobalScopeCaches.cpp:
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp: Removed.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h: Removed.
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::cacheStorageConnection):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::WebCacheStorageConnection):
(WebKit::WebCacheStorageConnection::~WebCacheStorageConnection):
(WebKit::WebCacheStorageConnection::connection):
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::remove):
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::lockStorage):
(WebKit::WebCacheStorageConnection::unlockStorage):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):
(WebKit::WebCacheStorageConnection::updateQuotaBasedOnSpaceUsage):
(WebKit::WebCacheStorageConnection::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageProvider.cpp:
(WebKit::WebCacheStorageProvider::createCacheStorageConnection):

Canonical link: <a href="https://commits.webkit.org/279805@main">https://commits.webkit.org/279805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/add67367a527c9683911aa59c1790f7871eaeb78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54611 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57892 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56913 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41571 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44234 "Passed tests") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3602 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56706 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32175 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25360 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28971 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4635 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3485 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59482 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29847 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51655 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47390 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51036 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11953 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31974 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30782 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
```diff
@@ -40,44 +40,43 @@ namespace WebKit {
+    RefPtr<IPC::Connection> m_connection;
+    Lock m_connectedIdentifierCountersLock;
+    HashCountedSet<WebCore::DOMCacheIdentifier> m_connectedIdentifierCounters WTF_GUARDED_BY_LOCK(m_connectedIdentifierCountersLock);
+    Lock m_clientOriginLockRequestCountersLock;
```
Can we just have one lock (e.g. `m_connectionLock`) and make all members protected by it since all operations require it now? Then we don't need to ensure the ordering of the locks

youennf replied:
Good point, it does not bring much to have 3 locks.

youennf replied:
Done

szewai commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -38,19 +38,33 @@
+        m_connection = connection;
+    }
+
+    return connection.releaseNonNull();
```
Is it safe to not hold lock between check and set? What will happen in the following case:
```
T1: hold lock -> check m_connection -> release lock
Main: create connection
Main: connection is closed, hold lock -> set m_connection null -> release lock
T1: hold lock -> set m_connection with connection -> release lock
```

youennf replied:
Right, let's only set m_connection in main thread to prevent this issue.

szewai commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
```diff
@@ -40,44 +40,41 @@ namespace WebKit {
+    Ref<IPC::Connection> connection();
+
+    Lock m_connectionLock;
+    RefPtr<IPC::Connection> m_connection;
```
Do we need to add `WTF_GUARDED_BY_LOCK` to `m_connection` too?

youennf replied:
Done

youennf committed:
Do not hit main thread when using CacheStorageConnection from a worker
https://bugs.webkit.org/show_bug.cgi?id=274967
rdar://129062867

Reviewed by Sihui Liu.

We make WebCacheStorageConnection thread safe to be called.
To do so, we add locks for getting the connection and for the hash maps.
To keep consistency between connection and hash maps, as was ensured before this patch, we lock both map lock and connection lock when updating the map.

We block the worker thread to get the connection and then we can use it directly from the worker.
In case of network process crash, we might have to block again the worker on main thread.

Covered by existing tests.

* Source/WebCore/Modules/cache/WindowOrWorkerGlobalScopeCaches.cpp:
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp: Removed.
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h: Removed.
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::cacheStorageConnection):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::WebCacheStorageConnection):
(WebKit::WebCacheStorageConnection::~WebCacheStorageConnection):
(WebKit::WebCacheStorageConnection::connection):
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::remove):
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::lockStorage):
(WebKit::WebCacheStorageConnection::unlockStorage):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):
(WebKit::WebCacheStorageConnection::updateQuotaBasedOnSpaceUsage):
(WebKit::WebCacheStorageConnection::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Source/WebKit/WebProcess/Cache/WebCacheStorageProvider.cpp:
(WebKit::WebCacheStorageProvider::createCacheStorageConnection):

Canonical link: https://commits.webkit.org/279805@main
